src/wayland/selection/data_device/mod.rs

```diff
@@ -113,84 +109,458 @@ use super::{
+                    )
+                    .unwrap();
+                let offer = WlDataOffer::from_id(dh, offer).unwrap();
+
+                // advertize the offer to the client
+                device.data_offer(&offer);
+
+                for mime_type in metadata.mime_types.iter().cloned() {
+                    offer.offer(mime_type);
+                }
+                offer.source_actions(DndAction::convert_slice(&metadata.dnd_actions));
+
+                device.enter((*serial).into(), self, location.x, location.y, Some(&offer));
+
+                offers.push(offer);
+            }
+
+            Some(WlOfferData {
+                state: offer_state,
+                source,
+                wl_offers: offers,
+                last_source_actions: metadata.dnd_actions,
+            })
+        } else {
+            None
+        }
+    }
+
+    fn motion<S: Source>(
+        &self,
+        _data: &mut D,
+        offer: Option<&mut WlOfferData<S>>,
+        seat: &Seat<D>,
+        location: Point<f64, Logical>,
+        time: u32,
+    ) {
+        let seat_data = seat
+            .user_data()
+            .get::<RefCell<SeatData<D::SelectionUserData>>>()
+            .unwrap()
+            .borrow();
+
+        if let Some(offer) = offer {
+            if let Some(new_metadata) = offer.source.metadata() {
+                if offer.last_source_actions != new_metadata.dnd_actions {
+                    for wl_offer in &offer.wl_offers {
+                        wl_offer.source_actions(DndAction::convert_slice(&new_metadata.dnd_actions));
+                    }
+
+                    offer.last_source_actions = new_metadata.dnd_actions;
+                }
+            };
+        }
+
+        for device in seat_data.known_data_devices() {
+            if device.id().same_client_as(&self.id()) {
+                device.motion(time, location.x, location.y);
+            }
+        }
+    }
+
+    fn leave<S: Source>(&self, _data: &mut D, offer: Option<&mut WlOfferData<S>>, seat: &Seat<D>) {
+        if let Some(offer) = offer {
+            if offer.state.lock().unwrap().dropped {
+                return;
+            }
+        }
+        let seat_data = seat
+            .user_data()
+            .get::<RefCell<SeatData<D::SelectionUserData>>>()
+            .unwrap()
+            .borrow();
+        for device in seat_data.known_data_devices() {
+            if device.id().same_client_as(&self.id()) {
+                device.leave();
+            }
+        }
+    }
+
+    fn drop<S: Source>(&self, _data: &mut D, offer: Option<&mut WlOfferData<S>>, seat: &Seat<D>) {
+        let seat_data = seat
+            .user_data()
+            .get::<RefCell<SeatData<D::SelectionUserData>>>()
+            .unwrap()
+            .borrow();
+        for device in seat_data.known_data_devices() {
+            if let Some(ref offer) = offer {
+                if device.version() >= 3 && offer.state.lock().unwrap().chosen_action.is_empty() {
```

Comment:
Is there a const available for this check?

Reply:
Kinda. `chosen_action` can only be set, if `SetActions` is available. I thus used that request's constant.